src/EnergyPlus/Pipes.hh

```diff
@@ -1,129 +1,53 @@
 		int BranchIndex; // Index of plant Branch index where this pipe resides
 		int CompIndex; // Index of plant Comp index where this pipe resides
-		bool OneTimeInit;
-		bool CheckEquipName;
-		bool EnvrnFlag;
 
 		// Default Constructor
 		LocalPipeData() :
-			TypeOf( 0 ),
 			InletNodeNum( 0 ),
 			OutletNodeNum( 0 ),
 			LoopNum( 0 ),
 			LoopSide( 0 ),
 			BranchIndex( 0 ),
-			CompIndex( 0 ),
-			OneTimeInit( true ),
-			CheckEquipName( true ),
-			EnvrnFlag( true )
+			CompIndex( 0 )
 		{}
 
-		// Member Constructor
-		LocalPipeData(
-			std::string const & Name, // main plant (cooling) loop ID
-			int const TypeOf, // type of pipe
-			int const InletNodeNum, // Node number on the inlet side of the plant
-			int const OutletNodeNum, // Node number on the inlet side of the plant
-			int const LoopNum, // Index of plant loop where this pipe resides
-			int const LoopSide, // Index of plant loop side where this pipe resides
-			int const BranchIndex, // Index of plant Branch index where this pipe resides
-			int const CompIndex, // Index of plant Comp index where this pipe resides
-			bool const OneTimeInit,
-			bool const CheckEquipName,
-			bool const EnvrnFlag
-		) :
-			Name( Name ),
-			TypeOf( TypeOf ),
-			InletNodeNum( InletNodeNum ),
-			OutletNodeNum( OutletNodeNum ),
-			LoopNum( LoopNum ),
-			LoopSide( LoopSide ),
-			BranchIndex( BranchIndex ),
-			CompIndex( CompIndex ),
-			OneTimeInit( OneTimeInit ),
-			CheckEquipName( CheckEquipName ),
-			EnvrnFlag( EnvrnFlag )
-		{}
+		static std::shared_ptr<PlantComponent> pipeFactory( std::string objectName );
+		int performEveryTimeInit();
+		int performOneTimeInit();
+		int performBeginEnvrnInit();
+		int simulate();
 
 	};
 
 	// Object Data
-	extern Array1D< LocalPipeData > LocalPipe; // dimension to number of pipes
-
-	// Functions
-
-	void
-	SimPipes(
-		int const CompType,
-		std::string & PipeName,
-		int & CompIndex,
-		Real64 const MaxVolFlowRate,
-		bool const InitLoopEquip,
-		bool const FirstHVACIteration
-	);
-
-	// End Plant Loop Module Driver Subroutines
-	//******************************************************************************
-
-	// Beginning of Plant Loop Module Get Input subroutines
-	//******************************************************************************
-
-	void
-	GetPipeInput();
-
-	// End of Get Input subroutines for the Plant Loop Module
-	//******************************************************************************
-
-	// Beginning Initialization Section of the Plant Loop Module
-	//******************************************************************************
-
-	void
-	InitializePipes(
-		int const PipeType, // Type of Pipe
-		std::string const & PipeName, // Name of Pipe
-		int & PipeNum, // Index into pipe structure for name
-		Real64 const MaxVolFlowRate // unused at present time
-	);
+	extern Array1D< std::shared_ptr< LocalPipeData > > LocalPipe; // dimension to number of pipes
```

Comment:
Well played, @Myoldmopar, well played. Outdate the diff so I can't find the conversation...

@DeadParrot Sorry, but I don't agree with much of that. I agree that for computations, indexing is often the way to go, but this is not that case. I would be fine with a rule for new code, but I wouldn't phrase it in terms of indexing. The rule should be standard containers for new code. Standard containers are, well, standard, and there's a wealth of information available on how to use them. Using special wrappers might help current developers, but presents a barrier to new developers. In my opinion that is a major net loss. Here are my responses to your list:

* I'm not sure if I understand your point, but why should containers of values work the same as containers of pointers? Seems to me that they're different.
* Yes, but that's OK, C++ is a bit of a mess
* Comparison function?!?
* Bounds checking only matters if there's a lot of indexing going on and I think I've been clear what I think about that.
* If they're using standard containers, then yes, that's something that Google can solve. I'd prefer a function over an entire object that'll require developers to sift through code to determine what it does.

Apologies for rejecting everything you're saying, but I don't see this as an average vs expert thing. In any case, I'd be happy to look at a design doc and code.